File: lib/java/src/main/java/uapi/internal/types/UStruct.java
package uapi.internal.types;

import static uapi.internal.generation.GenerateRandomStruct.generateRandomStruct;
import static uapi.internal.validation.ValidateStruct.validateStruct;

import java.util.List;
import java.util.Map;

import uapi.internal.generation.GenerateContext;
import uapi.internal.validation.ValidateContext;
import uapi.internal.validation.ValidationFailure;

public class UStruct implements UType {

    public static final String _STRUCT_NAME = "Object";

    public final String name;
    public final Map<String, UFieldDeclaration> fields;

    public UStruct(String name, Map<String, UFieldDeclaration> fields) {
        this.name = name;
        this.fields = fields;
    }

    @Override
    public int getTypeParameterCount() {
        return 0;
    }

    @Override
    public List<ValidationFailure> validate(Object value, List<UTypeDeclaration> typeParameters, ValidateContext ctx) {
        return validateStruct(value, this.name, this.fields, ctx);
    }

    @Override
    public Object generateRandomValue(GenerateContext ctx) {
        return generateRandomStruct(this.fields, ctx);
    }

    @Override
    public String getName() {
        return _STRUCT_NAME;
    }
}

File: lib/py/uapi/internal/types/UStruct.py
from typing import TYPE_CHECKING


if TYPE_CHECKING:
    from uapi.internal.validation.ValidateContext import ValidateContext
    from uapi.RandomGenerator import RandomGenerator
    from uapi.internal.validation.ValidationFailure import ValidationFailure
    from uapi.internal.types.UFieldDeclaration import UFieldDeclaration
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration
    from uapi.internal.generation.GenerateContext import GenerateContext

from uapi.internal.types.UType import UType

_STRUCT_NAME: str = "Object"


class UStruct(UType):

    def __init__(self, name: str, fields: dict[str, 'UFieldDeclaration']) -> None:
        self.name = name
        self.fields = fields

    def get_type_parameter_count(self) -> int:
        return 0

    def validate(self, value: object,
                 type_parameters: list['UTypeDeclaration'], ctx: 'ValidateContext') -> list['ValidationFailure']:
        from uapi.internal.validation.ValidateStruct import validate_struct
        return validate_struct(value, self.name, self.fields, ctx)

    def generate_random_value(self, ctx: 'GenerateContext') -> object:
        from uapi.internal.generation.GenerateRandomStruct import generate_random_struct
        return generate_random_struct(self.fields, ctx)

    def get_name(self) -> str:
        return _STRUCT_NAME
